### ISSUE TITLE
One peer in docker gets crash when load test

### ISSUE THREAD

**Issue by samding01** [hyperledger-archives/fabric#612]
A peer run in docker container gets crash when getting invoke/query by REST on Z system.
The replication steps is as follows:

1) Following #587, set the logging.level =error for chaincode_example02;
2) Start peer in docker:

> docker run -p 5000:5000 --rm -it -e OPENCHAIN_VM_ENDPOINT=http://172.17.0.1:4243 -e OPENCHAIN_PEER_ID=vp1 -e OPENCHAIN_PEER_ADDRESSAUTODETECT=true openchain-peer obc-peer peer

3) Delopy example02:

> OPENCHAIN_PEER_ADDRESS=172.17.0.2:30303 ./obc-peer chaincode deploy -p github.com/openblockchain/obc-peer/openchain/example/chaincode/chaincode_example02 -c '{"Function":"init", "Args": ["a","100", "b", "200"]}'

4) Using jmeter to send "invoke" and "query" REST requests in 20 thread, loop

> { "chaincodeSpec": { "type": "GOLANG", "chaincodeID": { "name":    <name>}, "ctorMsg": { "function":"invoke", "args": [ "a", "b", "1" ] } }    

and

> {"chaincodeSpec":{"type": "GOLANG","chaincodeID":{"name": <name> },"ctorMsg":{
> "function":"query","args":["a"] } } 

Then the obc-peer gets crash in less than 1 minutes. Here is part of msg:

>   ....
>  15:28:08.713 [peer] GetLocalAddress -> INFO 6674^[[0m Auto detected peer address: 172.17.0.2:30303
> 15:28:08.714 [consensus/controller] NewConsenter -> INFO 6675^[[0m Running with default consensus plugin (noops)
> fatal error: concurrent map read and map write
> goroutine 73010 [running]:
> runtime.throw(0x80baa440, 0x21)
>         /usr/local/go/src/runtime/panic.go:530 +0xb6 fp=0xc85baa4b20 sp=0xc85baa4b08
> runtime.mapaccess1_faststr(0x8089e2c0, 0xc82024cc30, 0xc85bd75e00, 0x24, 0x80183256)
>         /usr/local/go/src/runtime/hashmap_fast.go:202 +0x92 fp=0xc85baa4b80 sp=0xc85baa4b20
> github.com/openblockchain/obc-peer/openchain/chaincode.(_Handler).enterBusyState.func1(0xc85bc9bdc0, 0xc820142f00, 0x80af6aa0, 0x8)
>         /go/src/github.com/openblockchain/obc-peer/openchain/chaincode/handler.go:681 +0xda fp=0xc85baa4f88 sp=0xc85baa4b80
> runtime.goexit()
>         /usr/local/go/src/runtime/asm_s390x.s:1011 +0x2 fp=0xc85baa4f88 sp=0xc85baa4f88
> created by github.com/openblockchain/obc-peer/openchain/chaincode.(_Handler).enterBusyState
>         /go/src/github.com/openblockchain/obc-peer/openchain/chaincode/handler.go:788 +0x76
> ....

However there is no crash when running without using docker

**Comment by binhn** [comment]
@srderson could you investigate.

**Comment by srderson** [comment]
@samding01 I've submitted PR #620 if you'd like to grab it and see if it resolves the issue for you. Will wait for @muralisrini to review the fix.

**Comment by srderson** [comment]
@samding01 - @muralisrini was already able to review and accept the PR. We did not reproduce your issue, but based on the stack trace there was an obvious fix that needed to be made. Please catch up and test again. Once you have confirmed the problem is resolved, we'll close the issue.

**Comment by muralisrini** [comment]
Will just note that we did run concurrent queries and transactions without debug and did not see this particular crash. It is quite possible the Z environment is bringing out these issues. Great value running in z/OS and Power.

@samding01 : how are the invoke and query distributed in the loop test ?

**Comment by samding01** [comment]
@muralisrini : invoke and query are 1:1 sequentially in the loop.
Thanks,

**Comment by samding01** [comment]
Get the latest fixing code from master, and found it crash again:

> 15:43:00.229 [consensus/controller] NewConsenter -> INFO 12bd5^[[0m Running with default consensus plugin (noops)
> 15:43:00.238 [rest] Query -> INFO 12bd6^[[0m REST querying chaincode...
> 15:43:00.238 [peer] GetLocalAddress -> INFO 12bd7^[[0m Auto detected peer address: 172.17.0.2:30303
> 15:43:00.239 [consensus/controller] NewConsenter -> INFO 12bd8^[[0m Running with default consensus plugin (noops)
> fatal error: concurrent map read and map write
> goroutine 854715 [running]:
> runtime.throw(0x80bb62a0, 0x21)
>         /usr/local/go/src/runtime/panic.go:530 +0xb6 fp=0xc8b85bfc50 sp=0xc8b85bfc38
> runtime.mapaccess1_faststr(0x808a55c0, 0xc87d9f4e70, 0xc835947679, 0x1, 0xc87d401818)
>         /usr/local/go/src/runtime/hashmap_fast.go:202 +0x92 fp=0xc8b85bfcb0 sp=0xc8b85bfc50
> github.com/openblockchain/obc-peer/openchain/ledger/statemgmt/state.(_State).Get(0xc82014fb80, 0xc8b0672580, 0x80, 0xc835947679, 0x1, 0x1359475d9, 0x0, 0x0, 0x0, 0x0, ...)
>         /go/src/github.com/openblockchain/obc-peer/openchain/ledger/statemgmt/state/state.go:122 +0x13a fp=0xc8b85bfd88 sp=0xc8b85bfcb0
> github.com/openblockchain/obc-peer/openchain/ledger.(_Ledger).GetState(0xc82020f920, 0xc8b0672580, 0x80, 0xc835947679, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0, ...)
>         /go/src/github.com/openblockchain/obc-peer/openchain/ledger/ledger.go:191 +0x94 fp=0xc8b85bfde8 sp=0xc8b85bfd88
> github.com/openblockchain/obc-peer/openchain/chaincode.(_Handler).handleGetState.func1(0xc8b0672500, 0xc8a0aa7780)
>         /go/src/github.com/openblockchain/obc-peer/openchain/chaincode/handler.go:559 +0x4be fp=0xc8b85bffa8 sp=0xc8b85bfde8
> runtime.goexit()
>         /usr/local/go/src/runtime/asm_s390x.s:1011 +0x2 fp=0xc8b85bffa8 sp=0xc8b85bffa8
> created by github.com/openblockchain/obc-peer/openchain/chaincode.(_Handler).handleGetState
>         /go/src/github.com/openblockchain/obc-peer/openchain/chaincode/handler.go:580 +0x5e
> ....

By the way,  has the  #542 fixing code been checked into master branch?

**Comment by samding01** [comment]
Test on X host, there is no crash for the same scenario, the input REST requests get paused for a while but not stopped  when the memory is used up.

**Comment by muralisrini** [comment]
@samding01 :  #542 should be fixed. Note that though as invoke is asynch, it may not get executed in time for an immediate query to catch the update. However the invoke should get executed and when it does, the query should reflect it.
A request : would you mind running a scenario with just 1 thread running the invoke and query and other 19 threads running just the query please ?

**Comment by samding01** [comment]
@muralisrini :  Getting the latest code of the  master, running for 5 minutes for 20 threads of invoke and query,  no crash found.

However, like in #542, if we have a loop in the invoke function of the example02, the same scenario gets the crash with the same reason `fatal error: concurrent map read and map write`

```
 .....
// Write the state back to the ledger
err = stub.PutState(A, []byte(strconv.Itoa(Aval)))
if err != nil {
    return nil, err
}

err = stub.PutState(B, []byte(strconv.Itoa(Bval)))
if err != nil {
    return nil, err
}

for i := 0; i < 1000; i++ {
    err = stub.PutState(fmt.Sprintf("%c%d", A, i), []byte(strconv.Itoa(Aval)))
   if err != nil {
       return nil, err
   }

   err = stub.PutState(fmt.Sprintf("%c%d", B, i), []byte(strconv.Itoa(Bval)))
   if err != nil {
       return nil, err
   }
}

return nil, nil
```

}

**Comment by samding01** [comment]
Tested with the suggested scenario (1 thread for invoke and 19 threads for query). 
After a few seconds from starting, the "invoke" request stops (no response), and "query" returns the same value (eg. "a"=84, initial value =100).
 No crash. The cpu  load is very small. 
Seem it is  wait/lock internally.

**Comment by binhn** [comment]
Try with PR #625.  It will fix all these problems with synchronization of invokes.  @muralisrini please pull when you have a chance

**Comment by srderson** [comment]
Also PR #631 should resolve the query synchronization issue which is the cause of the stack trace posted a few comments above.

**Comment by binhn** [comment]
all related PRs are merged into master now
please try the master branch

**Comment by samding01** [comment]
Tested the latest master code on Z.  The crash is not found and the invoke requests are paused repeatedly.

> 14:57:44.762 [rest] Query -> INFO 12a6 REST querying chaincode...
> 14:57:44.768 [rest] Invoke -> INFO 12a7 Successfuly invoked chainCode with txuuid (d8491d87-430b-4d6f-a526-2c8c9bbd8340)
> 14:57:44.769 [rest] Query -> INFO 12a8 REST querying chaincode...
> 14:57:44.770 [rest] Invoke -> INFO 12a9 Successfuly invoked chainCode with txuuid (6c2e2adb-19c0-4390-a92a-657e6d4d7ba0)
> 14:57:44.771 [rest] Query -> INFO 12aa REST querying chaincode...
> 14:57:44.772 [rest] Query -> INFO 12ab REST querying chaincode...
> 14:57:44.776 [rest] Invoke -> INFO 12ac REST invoking chaincode...

The peer seems to postpone new invoke while  processing the accepted invokes. 
If this is the expected behavior, please close this issue.

**Comment by srderson** [comment]
@samding01 That is the expected behavior. NoOps consensus processes transactions serially so that the system remains deterministic. Queries transactions may run in parallel while invoke transactions are being processed. Thanks for your help!
